Shared/configurations/ConfigurationController.cpp

```diff
@@ -64,77 +79,308 @@ void ConfigurationController::select(int index)
-    connect(m_networkReply, &QNetworkReply::finished, this, &ConfigurationController::finished);
-    connect(m_networkReply, &QNetworkReply::errorOccurred, this, &ConfigurationController::downloadErrorOccurred);
+    m_configurationListModel->setDataByName(configurationName, bytesReceived * 1.0 / bytesTotal * 1.0 * 100, ConfigurationListModel::PercentDownloaded);
+  });
+  connect(contentReply, &QNetworkReply::readyRead, this, [this, contentReply, downloadFilePath, configurationName]()
+  {
+    readyRead(contentReply, downloadFilePath, configurationName);
+  });
+  connect(contentReply, &QNetworkReply::finished, this, [this, contentReply, downloadFilePath, configurationName]()
+  {
+    finished(contentReply, downloadFilePath, configurationName);
+  });
+  connect(contentReply, &QNetworkReply::errorOccurred, this, [this, contentReply](QNetworkReply::NetworkError error)
+  {
+    contentReply->abort();
+    if (error != QNetworkReply::NetworkError::OperationCanceledError)
+      emit toolErrorOccurred(QString("Network Error (%1)").arg(contentReply->errorString()), QStringLiteral("Error Downloading"));
   });
 }
 
-void ConfigurationController::cancel(int index)
+void ConfigurationController::download(int index)
 {
-  Q_UNUSED(index); // TODO: supporting multiple downloads at a time will require additional management
-  if (m_networkReply && m_networkReply->isOpen())
+  // fetch a copy of the selected model item from the listview
+  const auto configuration = m_configurationListModel->at(index);
+  const auto configurationUrl = configuration.url();
+  const auto configurationUrlStr = configuration.urlStr();
+  const auto configurationName = configuration.name();
+
+#ifdef Q_OS_ANDROID
+  if (configurationUrl.scheme() == QStringLiteral("content"))
   {
-    // cancel the downloading item and reset all the
-    m_configurationListModel->cancel(index);
-    m_aborted = true;
-    m_downloadInProgress = false;
-    m_networkReply->abort();
+    // copy file from the content provider as if it was a downloaded file
+    QString contentDownloadStr{generateUniqueDownloadFilePath()};
+    QtConcurrent::run([this, configurationUrlStr, configurationName, contentDownloadStr]
+    {
+      QFile fileContent{configurationUrlStr};
+      const auto bytesTotal = fileContent.size();
+      fileContent.open(QIODevice::ReadOnly);
+
+      QFile fileDownload{contentDownloadStr};
+      fileDownload.open(QIODevice::WriteOnly);
+      quint64 bytesWritten = 0;
+      quint8 percentTotal = 0;
+      constexpr qint64 sizeLine = 1024 * 16;
+
+      // write the lines to the temporary 'download' file and update the
+      // percentage downloaded by whole integers only to limit updates to the UI
+      while (!fileContent.atEnd())
+      {
+        bytesWritten += fileDownload.write(fileContent.readLine(sizeLine));
+        const quint8 percentLine = bytesWritten * 1.0 / bytesTotal * 100.0;
```

Comment:
```suggestion
        const quint8 percentLine = bytesWritten / bytesTotal * 100.0;
```
I assume '* 1' can be removed.

Reply:
Had this in there to ensure nothing caused integer division to happen. I think the ` * 100.0` at the end actually takes care of this already, yes.

Reply:
Actually this _was_ necessary to prevent integer division from happening when evaluating left to right. `bytesWritten` is integer type. I experimented with just using a float for bytesWritten instead here but the disassembly looked simpler with multiplying an integer type by 1.0 (at least on MacOS) to get the desired effect.